File: java/PrimeFinder.java
// PrimeFinder.java
import java.util.*;
import java.util.concurrent.*;
import java.util.stream.*;
import java.io.*;
import com.google.gson.Gson;
import com.google.gson.GsonBuilder;

public class PrimeFinder {
    
    // Check if a number is prime
    public static boolean isPrime(int n) {
        if (n <= 1) return false;
        if (n <= 3) return true;
        if (n % 2 == 0 || n % 3 == 0) return false;
        
        int i = 5;
        while (i * i <= n) {
            if (n % i == 0 || n % (i + 2) == 0) return false;
            i += 6;
        }
        return true;
    }
    
    // Find primes in a range
    public static List<Integer> findPrimesInRange(int start, int end) {
        List<Integer> primes = new ArrayList<>();
        for (int i = start; i <= end; i++) {
            if (isPrime(i)) {
                primes.add(i);
            }
        }
        return primes;
    }
    
    // Sequential implementation
    public static Result findPrimesSequential(int start, int end) {
        long startTime = System.nanoTime();
        List<Integer> primes = findPrimesInRange(start, end);
        long duration = System.nanoTime() - startTime;
        
        return new Result(start, end, primes.size(), duration / 1_000_000_000.0, 1, primes);
    }
    
    // Thread pool implementation
    public static Result findPrimesThreadPool(int start, int end, int numThreads) throws InterruptedException, ExecutionException {
        long startTime = System.nanoTime();
        
        ExecutorService executor = Executors.newFixedThreadPool(numThreads);
        List<Future<List<Integer>>> futures = new ArrayList<>();
        
        int chunkSize = (end - start + 1) / numThreads;
        if (chunkSize < 1) chunkSize = 1;
        
        // Submit tasks
        for (int i = start; i <= end; i += chunkSize) {
            final int rangeStart = i;
            final int rangeEnd = Math.min(i + chunkSize - 1, end);
            
            Future<List<Integer>> future = executor.submit(() -> findPrimesInRange(rangeStart, rangeEnd));
            futures.add(future);
        }
        
        // Collect results
        List<Integer> allPrimes = new ArrayList<>();
        for (Future<List<Integer>> future : futures) {
            allPrimes.addAll(future.get());
        }
        
        executor.shutdown();
        long duration = System.nanoTime() - startTime;
        
        Collections.sort(allPrimes);
        return new Result(start, end, allPrimes.size(), duration / 1_000_000_000.0, numThreads, allPrimes);
    }
    
    // CompletableFuture implementation
    public static Result findPrimesCompletableFuture(int start, int end, int numThreads) {
        long startTime = System.nanoTime();
        
        ExecutorService executor = Executors.newFixedThreadPool(numThreads);
        
        int chunkSize = (end - start + 1) / numThreads;
        if (chunkSize < 1) chunkSize = 1;
        
        List<CompletableFuture<List<Integer>>> futures = new ArrayList<>();
        
        // Create CompletableFutures
        for (int i = start; i <= end; i += chunkSize) {
            final int rangeStart = i;
            final int rangeEnd = Math.min(i + chunkSize - 1, end);
            
            CompletableFuture<List<Integer>> future = CompletableFuture.supplyAsync(
                () -> findPrimesInRange(rangeStart, rangeEnd),
                executor
            );
            futures.add(future);
        }
        
        // Combine all futures
        CompletableFuture<List<Integer>> combinedFuture = CompletableFuture
            .allOf(futures.toArray(new CompletableFuture[0]))
            .thenApply(v -> futures.stream()
                .map(CompletableFuture::join)
                .flatMap(List::stream)
                .sorted()
                .collect(Collectors.toList())
            );
        
        List<Integer> allPrimes = combinedFuture.join();
        executor.shutdown();
        
        long duration = System.nanoTime() - startTime;
        return new Result(start, end, allPrimes.size(), duration / 1_000_000_000.0, numThreads, allPrimes);
    }
    
    // Parallel Stream implementation
    public static Result findPrimesParallelStream(int start, int end) {
        long startTime = System.nanoTime();
        
        List<Integer> primes = IntStream.rangeClosed(start, end)
            .parallel()
            .filter(PrimeFinder::isPrime)
            .boxed()
            .collect(Collectors.toList());
        
        long duration = System.nanoTime() - startTime;
        Collections.sort(primes);
        
        return new Result(start, end, primes.size(), duration / 1_000_000_000.0, 
                         Runtime.getRuntime().availableProcessors(), primes);
    }
    
    public static void main(String[] args) throws Exception {
        // Parse command line arguments
        int start = 1;
        int end = 100000;
        int workers = Runtime.getRuntime().availableProcessors();
        String method = "all";
        boolean savePrimes = false;
        String outputFile = "java_results.json";
        
        for (int i = 0; i < args.length; i++) {
            switch (args[i]) {
                case "--start":
                    start = Integer.parseInt(args[++i]);
                    break;
                case "--end":
                    end = Integer.parseInt(args[++i]);
                    break;
                case "--workers":
                    workers = Integer.parseInt(args[++i]);
                    break;
                case "--method":
                    method = args[++i];
                    break;
                case "--save-primes":
                    savePrimes = true;
                    break;
                case "--output":
                    outputFile = args[++i];
                    break;
            }
        }
        
        System.out.println("Finding primes from " + start + " to " + end);
        System.out.println("Available processors: " + Runtime.getRuntime().availableProcessors());
        
        Map<String, Object> results = new HashMap<>();
        results.put("configuration", Map.of(
            "start_range", start,
            "end_range", end,
            "cpu_count", Runtime.getRuntime().availableProcessors()
        ));
        
        List<Integer> primes = null;
        
        if (method.equals("all") || method.equals("sequential")) {
            System.out.println("\nRunning sequential version...");
            Result seqResult = findPrimesSequential(start, end);
            if (!savePrimes) seqResult.primes = null;
            results.put("sequential", seqResult.toMap());
            primes = seqResult.primes;
            System.out.println("Sequential: " + seqResult.primesFound + " primes in " + 
                             String.format("%.4f", seqResult.executionTime) + " seconds");
        }
        
        if (method.equals("all") || method.equals("threadpool")) {
            System.out.println("\nRunning thread pool version with " + workers + " threads...");
            Result threadResult = findPrimesThreadPool(start, end, workers);
            if (!savePrimes) threadResult.primes = null;
            results.put("threadpool", threadResult.toMap());
            if (primes == null) primes = threadResult.primes;
            System.out.println("Thread Pool: " + threadResult.primesFound + " primes in " + 
                             String.format("%.4f", threadResult.executionTime) + " seconds");
        }
        
        if (method.equals("all") || method.equals("completable")) {
            System.out.println("\nRunning CompletableFuture version with " + workers + " threads...");
            Result cfResult = findPrimesCompletableFuture(start, end, workers);
            if (!savePrimes) cfResult.primes = null;
            results.put("completablefuture", cfResult.toMap());
            if (primes == null) primes = cfResult.primes;
            System.out.println("CompletableFuture: " + cfResult.primesFound + " primes in " + 
                             String.format("%.4f", cfResult.executionTime) + " seconds");
        }
        
        if (method.equals("all") || method.equals("parallel")) {
            System.out.println("\nRunning parallel stream version...");
            Result parallelResult = findPrimesParallelStream(start, end);
            if (!savePrimes) parallelResult.primes = null;
            results.put("parallelstream", parallelResult.toMap());
            if (primes == null) primes = parallelResult.primes;
            System.out.println("Parallel Stream: " + parallelResult.primesFound + " primes in " + 
                             String.format("%.4f", parallelResult.executionTime) + " seconds");
        }
        
        if (savePrimes && primes != null) {
            results.put("primes", primes);
        }
        
        // Save results to JSON
        Gson gson = new GsonBuilder().setPrettyPrinting().create();
        try (FileWriter writer = new FileWriter(outputFile)) {
            gson.toJson(results, writer);
        }
        
        System.out.println("\nResults saved to " + outputFile);
    }
    
    // Result class
    static class Result {
        int startRange;
        int endRange;
        int primesFound;
        double executionTime;
        int workers;
        List<Integer> primes;
        
        Result(int startRange, int endRange, int primesFound, double executionTime, 
               int workers, List<Integer> primes) {
            this.startRange = startRange;
            this.endRange = endRange;
            this.primesFound = primesFound;
            this.executionTime = executionTime;
            this.workers = workers;
            this.primes = primes;
        }
        
        Map<String, Object> toMap() {
            Map<String, Object> map = new HashMap<>();
            map.put("primes_found", primesFound);
            map.put("execution_time", executionTime);
            map.put("workers", workers);
            return map;
        }
    }
}

// PrimeFinderTest.java
import org.junit.Test;
import static org.junit.Assert.*;
import java.util.List;

public class PrimeFinderTest {
    
    @Test
    public void testIsPrime() {
        assertFalse(PrimeFinder.isPrime(1));
        assertTrue(PrimeFinder.isPrime(2));
        assertTrue(PrimeFinder.isPrime(3));
        assertFalse(PrimeFinder.isPrime(4));
        assertTrue(PrimeFinder.isPrime(5));
        assertFalse(PrimeFinder.isPrime(100));
        assertTrue(PrimeFinder.isPrime(101));
    }
    
    @Test
    public void testFindPrimesInRange() {
        List<Integer> primes = PrimeFinder.findPrimesInRange(1, 10);
        assertEquals(4, primes.size());
        assertTrue(primes.contains(2));
        assertTrue(primes.contains(3));
        assertTrue(primes.contains(5));
        assertTrue(primes.contains(7));
    }
    
    @Test
    public void testConcurrentResults() throws Exception {
        // Test that all methods find the same primes
        PrimeFinder.Result sequential = PrimeFinder.findPrimesSequential(1, 1000);
        PrimeFinder.Result threadPool = PrimeFinder.findPrimesThreadPool(1, 1000, 4);
        PrimeFinder.Result completable = PrimeFinder.findPrimesCompletableFuture(1, 1000, 4);
        
        assertEquals(sequential.primesFound, threadPool.primesFound);
        assertEquals(sequential.primesFound, completable.primesFound);
    }
}

File: python/prime_finder.py
# prime_finder.py
import argparse
import json
import time
import threading
import multiprocessing
from concurrent.futures import ThreadPoolExecutor, ProcessPoolExecutor
from typing import List, Tuple
import os

def is_prime(n: int) -> bool:
    """Check if a number is prime using trial division"""
    if n <= 1:
        return False
    if n <= 3:
        return True
    if n % 2 == 0 or n % 3 == 0:
        return False
    
    i = 5
    while i * i <= n:
        if n % i == 0 or n % (i + 2) == 0:
            return False
        i += 6
    return True

def find_primes_in_range(start: int, end: int) -> List[int]:
    """Find all primes in a given range"""
    return [i for i in range(start, end + 1) if is_prime(i)]

def find_primes_sequential(start: int, end: int) -> Tuple[List[int], float]:
    """Find primes sequentially"""
    start_time = time.time()
    primes = find_primes_in_range(start, end)
    duration = time.time() - start_time
    return primes, duration

def find_primes_threading(start: int, end: int, num_threads: int) -> Tuple[List[int], float]:
    """Find primes using threading"""
    start_time = time.time()
    
    chunk_size = (end - start + 1) // num_threads
    if chunk_size < 1:
        chunk_size = 1
    
    all_primes = []
    lock = threading.Lock()
    
    def worker(range_start, range_end):
        primes = find_primes_in_range(range_start, range_end)
        with lock:
            all_primes.extend(primes)
    
    with ThreadPoolExecutor(max_workers=num_threads) as executor:
        futures = []
        for i in range(start, end + 1, chunk_size):
            range_end = min(i + chunk_size - 1, end)
            future = executor.submit(worker, i, range_end)
            futures.append(future)
        
        # Wait for all to complete
        for future in futures:
            future.result()
    
    duration = time.time() - start_time
    return sorted(all_primes), duration

def find_primes_multiprocessing(start: int, end: int, num_processes: int) -> Tuple[List[int], float]:
    """Find primes using multiprocessing"""
    start_time = time.time()
    
    chunk_size = (end - start + 1) // num_processes
    if chunk_size < 1:
        chunk_size = 1
    
    # Create ranges for each process
    ranges = []
    for i in range(start, end + 1, chunk_size):
        range_end = min(i + chunk_size - 1, end)
        ranges.append((i, range_end))
    
    with ProcessPoolExecutor(max_workers=num_processes) as executor:
        # Map the function to all ranges
        results = executor.map(lambda r: find_primes_in_range(r[0], r[1]), ranges)
        
        # Flatten results
        all_primes = []
        for primes in results:
            all_primes.extend(primes)
    
    duration = time.time() - start_time
    return sorted(all_primes), duration

def benchmark_all_methods(start: int, end: int, workers: int):
    """Benchmark all methods and return results"""
    results = {}
    
    # Sequential
    print("Running sequential version...")
    primes, duration = find_primes_sequential(start, end)
    results['sequential'] = {
        'primes_found': len(primes),
        'execution_time': duration,
        'workers': 1
    }
    print(f"Sequential: {len(primes)} primes in {duration:.4f} seconds")
    
    # Threading
    print(f"\nRunning threading version with {workers} threads...")
    primes_thread, duration_thread = find_primes_threading(start, end, workers)
    results['threading'] = {
        'primes_found': len(primes_thread),
        'execution_time': duration_thread,
        'workers': workers,
        'speedup': duration / duration_thread
    }
    print(f"Threading: {len(primes_thread)} primes in {duration_thread:.4f} seconds")
    print(f"Speedup: {results['threading']['speedup']:.2f}x")
    
    # Multiprocessing
    print(f"\nRunning multiprocessing version with {workers} processes...")
    primes_multi, duration_multi = find_primes_multiprocessing(start, end, workers)
    results['multiprocessing'] = {
        'primes_found': len(primes_multi),
        'execution_time': duration_multi,
        'workers': workers,
        'speedup': duration / duration_multi
    }
    print(f"Multiprocessing: {len(primes_multi)} primes in {duration_multi:.4f} seconds")
    print(f"Speedup: {results['multiprocessing']['speedup']:.2f}x")
    
    return results, primes

def main():
    parser = argparse.ArgumentParser(description='Prime number finder with different concurrency methods')
    parser.add_argument('--start', type=int, default=1, help='Start of range')
    parser.add_argument('--end', type=int, default=100000, help='End of range')
    parser.add_argument('--workers', type=int, default=os.cpu_count(), help='Number of workers')
    parser.add_argument('--method', choices=['sequential', 'threading', 'multiprocessing', 'all'], 
                       default='all', help='Method to use')
    parser.add_argument('--save-primes', action='store_true', help='Save actual prime numbers')
    parser.add_argument('--output', default='python_results.json', help='Output file')
    
    args = parser.parse_args()
    
    print(f"Finding primes from {args.start} to {args.end}")
    print(f"CPU count: {os.cpu_count()}")
    
    if args.method == 'all':
        results, primes = benchmark_all_methods(args.start, args.end, args.workers)
        
        # Add configuration to results
        results['configuration'] = {
            'start_range': args.start,
            'end_range': args.end,
            'cpu_count': os.cpu_count()
        }
        
        if args.save_primes:
            results['primes'] = primes
    else:
        # Run specific method
        if args.method == 'sequential':
            primes, duration = find_primes_sequential(args.start, args.end)
            workers = 1
        elif args.method == 'threading':
            primes, duration = find_primes_threading(args.start, args.end, args.workers)
            workers = args.workers
        else:  # multiprocessing
            primes, duration = find_primes_multiprocessing(args.start, args.end, args.workers)
            workers = args.workers
        
        results = {
            args.method: {
                'primes_found': len(primes),
                'execution_time': duration,
                'workers': workers
            },
            'configuration': {
                'start_range': args.start,
                'end_range': args.end,
                'cpu_count': os.cpu_count()
            }
        }
        
        if args.save_primes:
            results['primes'] = primes
        
        print(f"\nFound {len(primes)} primes in {duration:.4f} seconds")
    
    # Save results
    with open(args.output, 'w') as f:
        json.dump(results, f, indent=2)
    
    print(f"\nResults saved to {args.output}")

if __name__ == "__main__":
    main()

# test_prime_finder.py
import unittest
from prime_finder import is_prime, find_primes_in_range

class TestPrimeFinder(unittest.TestCase):
    def test_is_prime(self):
        """Test the is_prime function"""
        test_cases = [
            (1, False),
            (2, True),
            (3, True),
            (4, False),
            (5, True),
            (100, False),
            (101, True),
            (1000, False),
            (1009, True),
        ]
        
        for n, expected in test_cases:
            with self.subTest(n=n):
                self.assertEqual(is_prime(n), expected)
    
    def test_find_primes_in_range(self):
        """Test finding primes in a range"""
        primes = find_primes_in_range(1, 10)
        self.assertEqual(primes, [2, 3, 5, 7])
        
        primes = find_primes_in_range(10, 20)
        self.assertEqual(primes, [11, 13, 17, 19])
    
    def test_empty_range(self):
        """Test empty range"""
        primes = find_primes_in_range(0, 1)
        self.assertEqual(primes, [])

if __name__ == '__main__':
    unittest.main()